loopd/daemon.go

```diff
@@ -119,117 +140,291 @@ func daemon(config *Config, lisCfg *listenerCfg) error {
 	// We'll also create and start an accompanying proxy to serve clients
 	// through REST.
 	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	d.restCtxCancel = cancel
 	mux := proxy.NewServeMux(customMarshalerOption)
 	var restHandler http.Handler = mux
-	if config.CORSOrigin != "" {
-		restHandler = allowCORS(restHandler, config.CORSOrigin)
+	if d.cfg.CORSOrigin != "" {
+		restHandler = allowCORS(restHandler, d.cfg.CORSOrigin)
 	}
 	proxyOpts := []grpc.DialOption{
 		grpc.WithInsecure(),
 		grpc.WithDefaultCallOptions(maxMsgRecvSize),
 	}
 	err = looprpc.RegisterSwapClientHandlerFromEndpoint(
-		ctx, mux, config.RPCListen, proxyOpts,
+		ctx, mux, d.cfg.RPCListen, proxyOpts,
 	)
 	if err != nil {
-		return err
+		return d.stopAndReturnErr(err)
 	}
 
-	restListener, err := lisCfg.restListener()
+	d.restListener, err = d.listenerCfg.restListener()
 	if err != nil {
-		return fmt.Errorf("REST proxy unable to listen on %s",
-			config.RESTListen)
+		return d.stopAndReturnErr(fmt.Errorf("REST proxy unable to "+
+			"listen on %s: %v", d.cfg.RESTListen, err))
 	}
 
 	// A nil listener indicates REST is disabled.
-	if restListener != nil {
+	if d.restListener != nil {
 		log.Infof("Starting REST proxy listener")
 
-		defer restListener.Close()
-		proxy := &http.Server{Handler: restHandler}
+		d.restServer = &http.Server{Handler: restHandler}
 
+		d.wg.Add(1)
 		go func() {
-			err := proxy.Serve(restListener)
+			defer d.wg.Done()
+
+			log.Infof("REST proxy listening on %s",
+				d.restListener.Addr())
+			err := d.restServer.Serve(d.restListener)
 			// ErrServerClosed is always returned when the proxy is
 			// shut down, so don't log it.
 			if err != nil && err != http.ErrServerClosed {
 				log.Error(err)
+				d.internalErrChan <- err
```

Comment:
The log statements in the goroutines (also below) can be moved to the main error handling goroutine. Maybe add a comment everywhere that the channel has enough buffer to not block.